src/nodejs-common/service.ts

```diff
@@ -0,0 +1,279 @@
+   * BigQuery or Storage.
+   *
+   * This handles making authenticated requests by exposing a `makeReq_`
+   * function.
+   *
+   * @constructor
+   * @alias module:common/service
+   *
+   * @param {object} config - Configuration object.
+   * @param {string} config.baseUrl - The base URL to make API requests to.
+   * @param {string[]} config.scopes - The scopes required for the request.
+   * @param {object=} options - [Configuration object](#/docs).
+   */
+  constructor(config: ServiceConfig, options: ServiceOptions = {}) {
+    this.baseUrl = config.baseUrl;
+    this.apiEndpoint = config.apiEndpoint;
+    this.timeout = options.timeout;
+    this.globalInterceptors = arrify(options.interceptors_!);
+    this.interceptors = [];
+    this.packageJson = config.packageJson;
+    this.projectId = options.projectId || PROJECT_ID_TOKEN;
+    this.projectIdRequired = config.projectIdRequired !== false;
+    this.providedUserAgent = options.userAgent;
+
+    const reqCfg = extend({}, config, {
+      projectIdRequired: this.projectIdRequired,
+      projectId: this.projectId,
+      authClient: options.authClient,
+      credentials: options.credentials,
+      keyFile: options.keyFilename,
+      email: options.email,
+      token: options.token,
+    });
+
+    this.makeAuthenticatedRequest =
+      util.makeAuthenticatedRequestFactory(reqCfg);
+    this.authClient = this.makeAuthenticatedRequest.authClient;
+    this.getCredentials = this.makeAuthenticatedRequest.getCredentials;
+
+    const isCloudFunctionEnv = !!process.env.FUNCTION_NAME;
+
+    if (isCloudFunctionEnv) {
+      this.interceptors.push({
+        request(reqOpts: DecorateRequestOptions) {
+          reqOpts.forever = false;
+          return reqOpts;
+        },
+      });
+    }
+  }
+
+  /**
+   * Return the user's custom request interceptors.
+   */
+  getRequestInterceptors(): Function[] {
+    // Interceptors should be returned in the order they were assigned.
+    return ([] as Interceptor[]).slice
+      .call(this.globalInterceptors)
+      .concat(this.interceptors)
+      .filter(interceptor => typeof interceptor.request === 'function')
+      .map(interceptor => interceptor.request);
+  }
+
+  /**
+   * Get and update the Service's project ID.
+   *
+   * @param {function} callback - The callback function.
+   */
+  getProjectId(): Promise<string>;
+  getProjectId(callback: (err: Error | null, projectId?: string) => void): void;
+  getProjectId(
+    callback?: (err: Error | null, projectId?: string) => void
+  ): Promise<string> | void {
+    if (!callback) {
+      return this.getProjectIdAsync();
+    }
+    this.getProjectIdAsync().then(p => callback(null, p), callback);
+  }
+
+  protected async getProjectIdAsync(): Promise<string> {
+    const projectId = await this.authClient.getProjectId();
+    if (this.projectId === PROJECT_ID_TOKEN && projectId) {
+      this.projectId = projectId;
+    }
+    return this.projectId;
+  }
+
```

Comment:
In the future, for functionality we internalize, we can use a `Promise`-only interface to reduce the API surface